# Web Mining and Applied NLP (44-620)

## Requests, JSON, and NLP

### Student Name: Cole Nollette

### Date: 11/12/24

### GitHub: https://github.com/nollettecs/Module4_44620

Perform the tasks described in the Markdown cells below.  When you have completed the assignment make sure your code cells have all been run (and have output beneath them) and ensure you have committed and pushed ALL of your changes to your assignment repository.

Make sure you have [installed spaCy and its pipeline](https://spacy.io/usage#quickstart) and [spaCyTextBlob](https://spacy.io/universe/project/spacy-textblob)

Every question that requires you to write code will have a code cell underneath it; you may either write your entire solution in that cell or write it in a python file (`.py`), then import and run the appropriate code to answer the question.

This assignment requires that you write additional files (either JSON or pickle files); make sure to submit those files in your repository as well.

# Installing Spacy and SpacyTextBlob

In [1]:
%pip install spacy
%pip install spacytextblob

  Using cached numpy-2.0.2-cp312-cp312-macosx_14_0_arm64.whl.metadata (60 kB)
Using cached numpy-2.0.2-cp312-cp312-macosx_14_0_arm64.whl (5.0 MB)
  Attempting uninstall: numpy
    Found existing installation: numpy 1.26.4
    Uninstalling numpy-1.26.4:
      Successfully uninstalled numpy-1.26.4
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
contourpy 1.2.0 requires numpy<2.0,>=1.20, but you have numpy 2.0.2 which is incompatible.
pyarrow 15.0.0 requires numpy<2,>=1.16.6, but you have numpy 2.0.2 which is incompatible.
matplotlib 3.8.2 requires numpy<2,>=1.21, but you have numpy 2.0.2 which is incompatible.
pandas 2.2.0 requires numpy<2,>=1.26.0; python_version >= "3.12", but you have numpy 2.0.2 which is incompatible.
scipy 1.12.0 requires numpy<1.29.0,>=1.22.4, but you have numpy 2.0.2 which is incompatible.
Note: you may need to restart the kernel to use up

# Question #1:
1. The following code accesses the [lyrics.ovh](https://lyricsovh.docs.apiary.io/#reference/0/lyrics-of-a-song/search) public api, searches for the lyrics of a song, and stores it in a dictionary object.  Write the resulting json to a file (either a JSON file or a pickle file; you choose). You will read in the contents of this file for future questions so we do not need to frequently access the API.

In [4]:
# Create and activate a Python virtual environment. 
# Before starting the project, try all these imports FIRST
# Address any errors you get running this code cell 
# by installing the necessary packages into your active Python environment.
# Try to resolve issues using your materials and the web.
# If that doesn't work, ask for help in the discussion forums.
# You can't complete the exercises until you import these - start early! 
# We also import json and pickle (included in the Python Standard Library).

import json
import pickle

import requests
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

print('All prereqs installed.')
!pip list

result = json.loads(requests.get('https://api.lyrics.ovh/v1/They Might Be Giants/Birdhouse in your soul').text)


All prereqs installed.
Package                   Version
------------------------- ---------------
annotated-types           0.7.0
anyio                     4.2.0
appnope                   0.1.3
argon2-cffi               23.1.0
argon2-cffi-bindings      21.2.0
arrow                     1.3.0
asttokens                 2.4.1
async-lru                 2.0.4
attrs                     23.2.0
Babel                     2.14.0
beautifulsoup4            4.12.3
bleach                    6.1.0
blis                      1.0.1
catalogue                 2.0.10
certifi                   2024.2.2
cffi                      1.16.0
charset-normalizer        3.3.2
click                     8.1.7
cloudpathlib              0.20.0
comm                      0.2.1
confection                0.1.5
contourpy                 1.2.0
cycler                    0.12.1
cymem                     2.0.8
debugpy                   1.8.0
decorator                 5.1.1
defusedxml                0.7.1
executing                

# Question 1b:

In [5]:
import requests
import json
import pickle

# Access the lyrics.ovh
response = requests.get('https://api.lyrics.ovh/v1/They%20Might%20Be%20Giants/Birdhouse%20in%20your%20soul')
lyrics_data = response.json()

# Write JSON to a file
with open('lyrics.json', 'w') as json_file:
    json.dump(lyrics_data, json_file, indent=4)
print("Lyrics saved to lyrics.json")

Lyrics saved to lyrics.json


# Question 2:
2. Read in the contents of your file.  Print the lyrics of the song (not the entire dictionary!) and use spaCyTextBlob to perform sentiment analysis on the lyrics.  Print the polarity score of the sentiment analysis.  Given that the range of the polarity score is `[-1.0,1.0]` which corresponds to how positive or negative the text in question is, do you think the lyrics have a more positive or negative connotaion?  Answer this question in a comment in your code cell.

In [3]:
!python3 -m spacy download en_core_web_sm

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.8/12.8 MB 26.3 MB/s eta 0:00:00a 0:00:01
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')


In [4]:
import json
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

# Load spaCy model
nlp_model = spacy.load('en_core_web_sm')

# Add sentiment analysis pipeline
nlp_model.add_pipe("spacytextblob", last=True)

# Open and load the JSON file
with open('lyrics.json', 'r') as file:
    song_data = json.load(file)

# Retrieve lyrics from dictionary
song_lyrics = song_data.get('lyrics', '')

# Perform sentiment analysis
doc_analysis = nlp_model(song_lyrics)
sentiment_score = doc_analysis._.blob.polarity  # Get the polarity

# Display the lyrics and the sentiment score
print("Song Lyrics:\n", song_lyrics)
print("\nSentiment Score:", sentiment_score)

# Interpret the sentiment
if sentiment_score > 0:
    print("The lyrics are positive.")
elif sentiment_score < 0:
    print("The lyrics are negative.")
else:
    print("The lyrics are neutral.")

Song Lyrics:
 I'm your only friend 
I'm not your only friend 
But I'm a little glowing friend 
But really I'm not actually your friend 
But I am 


Blue canary in the outlet by the light switch 

Who watches over you 

Make a little birdhouse in your soul 

Not to put too fine a point on it 

Say I'm the only bee in your bonnet 

Make a little birdhouse in your soul 



I have a secret to tell 

From my electrical well 

It's a simple message and I'm leaving out the whistles and bells 

So the room must listen to me 

Filibuster vigilantly 

My name is blue canary one note* spelled l-i-t-e 

My story's infinite 

Like the Longines Symphonette it doesn't rest 



Blue canary in the outlet by the light switch 

Who watches over you 

Make a little birdhouse in your soul 

Not to put too fine a point on it 

Say I'm the only bee in your bonnet 

Make a little birdhouse in your soul 



I'm your only friend 

I'm not your only friend 

But I'm a little glowing friend 

But really I'm not a

# Question 3:
3. Write a function that takes an artist, song, and filename, accesses the lyrics.ovh api to get the song lyrics, and writes the results to the specified filename.  Test this function by getting the lyrics to any four songs of your choice and storing them in different files.

In [6]:
import requests
import json

def save_song_lyrics(artist, song, filename):
   
    # Format the URL with the artist and song title
    url = f'https://api.lyrics.ovh/v1/{artist}/{song}'
    response = requests.get(url)
    
    if response.status_code == 200:
        # Parse the JSON response
        lyrics_data = response.json()
        
        # Write the lyrics to the specified file
        with open(filename, 'w') as file:
            json.dump(lyrics_data, file, indent=4)
        
        print(f"Lyrics for '{song}' by {artist} saved to {filename}")
    else:
        print(f"Failed to retrieve lyrics for '{song}' by {artist}. Status code: {response.status_code}")

# Testing the function with four songs
save_song_lyrics("The Beatles", "Let It Be", "let_it_be_lyrics.json")
save_song_lyrics("The Beatles", "Yesterday", "yesterday_lyrics.json")
save_song_lyrics("Queen", "Under Pressure", "under_pressure_lyrics.json")
save_song_lyrics("Queen", "Bohemian Rhapsody", "bohemian_rhapsody_lyrics.json")

Lyrics for 'Let It Be' by The Beatles saved to let_it_be_lyrics.json
Lyrics for 'Yesterday' by The Beatles saved to yesterday_lyrics.json
Lyrics for 'Under Pressure' by Queen saved to under_pressure_lyrics.json
Lyrics for 'Bohemian Rhapsody' by Queen saved to bohemian_rhapsody_lyrics.json


# Question 4:
4. Write a function that takes the name of a file that contains song lyrics, loads the file, performs sentiment analysis, and returns the polarity score.  Use this function to print the polarity scores (with the name of the song) of the three files you created in question 3.  Does the reported polarity match your understanding of the song's lyrics? Why or why not do you think that might be?  Answer the questions in either a comment in the code cell or a markdown cell under the code cell.

In [7]:
import json
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

# Load spaCy model and SpacyTextBlob
nlp = spacy.load("en_core_web_sm")
nlp.add_pipe("spacytextblob")

def analyze_lyrics_sentiment(filename, song_name):
    # Load the lyrics from JSON
    with open(filename, 'r') as file:
        lyrics_data = json.load(file)
    
    # Retrieve the lyrics text
    lyrics = lyrics_data.get('lyrics', '')
    
    # Perform sentiment analysis
    doc = nlp(lyrics)
    polarity_score = doc._.blob.polarity
    
    # Print the song name and its polarity score
    print(f"Polarity Score for '{song_name}': {polarity_score}")
    return polarity_score

# Test the function with 3 JSON files
analyze_lyrics_sentiment("let_it_be_lyrics.json", "Let It Be")
analyze_lyrics_sentiment("yesterday_lyrics.json", "Yesterday")
analyze_lyrics_sentiment("under_pressure_lyrics.json", "Under Pressure")

Polarity Score for 'Let It Be': 0.0801785714285714
Polarity Score for 'Yesterday': 0.010000000000000002
Polarity Score for 'Under Pressure': 0.34564814814814815


0.34564814814814815

# Question 4b: Polarity Scores
Does the reported polarity match your understanding of the song's lyrics? Why or why not do you think that might be?  Answer the questions in either a comment in the code cell or a markdown cell under the code cell.

Polarity scores represent the sentiment of a text on a scale ranging from -1 to 1:

Positive scores (closer to 1) indicate positive sentiment.
Negative scores (closer to -1) indicate negative sentiment.
Scores around 0 suggest a neutral sentiment.

"Let It Be" (0.0802): A score that's very close to 0. Meaning the lyrics are generally neutral with a slight lean towards positivity. This could reflect a tone of acceptance or calm. Overall I would say this seems to accurate as I wouldn't label this song as either positive or negative.

"Yesterday" (0.01): This score is also very close to 0. This indicates an almost neutral sentiment with a tiny positive inclination. This may suggest mixed emotions in the lyrics. I personally thought this song would have leaned more towards a negative score.  This sound tends to have more of a depressing tone to it as it reflects on the past being better than the future.

"Under Pressure" (0.3456): This has a higher positive score, slightly closer to 1 than the previous songs. This reflects more positivity or hope. Although the song discusses stress, it also emphasizes themes of resilience and community, which may contribute to a more positive sentiment. I would say I find this to be a pretty accurate score. 

Overall, the polarity scores produced what I would call accurate results.  

# Convert To HTML

In [8]:
!jupyter nbconvert --to html requests-json-nlp.ipynb

[NbConvertApp] Converting notebook requests-json-nlp.ipynb to html
[NbConvertApp] Writing 275570 bytes to requests-json-nlp.html
